### PR TITLE
GH-2318-mm: As a developer I want support for module declarations in d.ts

### PR DESCRIPTION
**Pull request by mmews-n4**
This is an alternative solution to the already merged solution of #2318. The major difference is that this solutions uses virtual resources in contrast to the already merged solution that models declared modules explicitly. The major benefit is that this solution does not need any changes of the N4JS AST or any of the language relevant code. Actually, the number of changed files outside of the `dts` bundle is reduced to 3. A drawback of this solution is that a new concept needs to be supported by the builder: i.e. virtual resources.

### CODE REVIEW

**Comment by mor-n4** [comment]
This looks like a good, elegant approach and in any case it's a clever idea to implement it that way!

However, the advantage of fewer changes to the AST model and types model come at the cost of a new, non-trivial feature in the core infrastructure (as you already said). So, while we would no longer have to deal with the differentiation between TModule and TDeclaredModule we would have to consider the special case of virtual resources in some (probably fewer) places.

Overall, I slightly favor the approach of the original PR, because it seems to represent more directly the concepts we are dealing with in case of module declarations and does not change the infrastructure. But I'd also be fine with this more concise approach.

Proposal: let's come back to this once we have a few more features of the dts-support ironed out, maybe in 2-3 weeks from now. (in any case, good to have this PR as reference!)

**Review by mor-n4 — APPROVED** [review]
After resolving the remaining issues: ship it! 🚂 

Also, please note the two points below...

File: plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSDeclaredNameValidator.xtend [hunk 1]
```diff
@@ -354,7 +355,7 @@ class N4JSDeclaredNameValidator extends AbstractN4JSDeclarativeValidator {
 									
 									if (!( // do not create issues for polyfills conflicting with imports, as they might fill them
 										dupeEO instanceof N4ClassifierDeclaration && baseEO instanceof ImportSpecifier &&
-										N4JSLanguageUtils.isNonStaticPolyfill(dupeEO as N4ClassifierDeclaration)
+										N4JSLanguageUtils.isNonStaticPolyfill(dupeEO as N4ClassDeclaration)
```
[the remark this answers is not in the document]
Comment (mmews-n4) [reply plugins/org.eclipse.n4js/src/org/eclipse/n4js/validation/validators/N4JSDeclaredNameValidator.xtend:358]: Thanks!

**Comment by mmews-n4** [comment]
Thank you for the review. As discussed, I will merge this now to master despite of a small problem with the incremental build, which I will continue to work on right away.

**Pull request merged**